core/src/main/java/jenkins/util/TimeProvider.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2019, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.util;
+
+import hudson.Extension;
+import hudson.ExtensionPoint;
+import jenkins.model.Jenkins;
+
+import javax.annotation.Nonnull;
+
+public interface TimeProvider extends ExtensionPoint {
```

Comment:
This is basically reimplementing [Clock](https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html).

Reply:
So are you proposing to create a ClockProvider that will be injectable using ExtensionPoint and to write a wrapper around the ManualTimeProvider ?

Otherwise I am not understanding your comment :)

Reply:
You could make a fake extension point that combines the two. I'm suggesting a standard API rather than reinventing one.

Reply:
Also, other plugins have mocked a Clock instance for use in time-dependent unit tests. Promoting this pattern might help encourage better unit testing (you have a lot of other great example unit tests for others to copy/paste, so this is a nice opportunity).

Reply:
Yeah, this could be improved, along some of the ideas Matt suggests.